sdk/dotnet/Pulumi/Deployment/TaskMonitoringHelper.cs

```diff
@@ -0,0 +1,191 @@
+// Copyright 2016-2021, Pulumi Corporation
+
+using System;
+using System.Linq;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Pulumi
+{
+    /// Monitors dynamically added tasks for completion. Enters IDLE
+    /// state when all monitored tasks finish. Allows awaiting next
+    /// IDLE state or an exception, whichever comes first.
+    /// Thread-safe.
+    internal sealed class TaskMonitoringHelper
+    {
+        private readonly TaskExceptionTracker errTracker = new TaskExceptionTracker();
+        private readonly TaskIdleTracker idleTracker = new TaskIdleTracker();
+
+        /// Starts monitoring the given task.
+        public void AddTask(Task task)
+        {
+            errTracker.AddTask(task);
+            idleTracker.AddTask(task);
+        }
+
+        /// Awaits next IDLE state or an exception, whichever comes
+        /// first. Several exceptions may be returned if they have
+        /// been observed prior to this call.
+        ///
+        /// IDLE state is represented as an empty sequence in the result.
+        public async Task<IEnumerable<Exception>> AwaitIdleOrFirstExceptionAsync()
+        {
+            var error = errTracker.AwaitExceptionAsync();
+            var idle = idleTracker.AwaitIdleAsync();
+            var first = await Task.WhenAny((Task)error, idle).ConfigureAwait(false);
+            if (first == idle)
+            {
+                return Enumerable.Empty<Exception>();
+            }
+            var err = await error;
```

Comment:
Nit. Also, I'd use `exceptionTask` instead of `error`.
```suggestion
            return await error;
```